File: src/types/conflicts.ts
/**
 * Conflict detection and resolution types.
 */

import { NodeId } from "./node.js";

/**
 * Conflict between two proposals.
 */
export interface ProposalConflict {
  /** The conflicting proposals */
  proposals: [string, string]; // Proposal IDs
  /** Nodes that both proposals modify */
  conflictingNodes: NodeId[];
  /** Fields that conflict (if field-level detection) */
  conflictingFields?: {
    [nodeId: string]: string[]; // Node ID -> field names
  };
  /** Severity of the conflict */
  severity: "field" | "node" | "critical";
  /** Whether conflict can be auto-resolved */
  autoResolvable: boolean;
}

/**
 * Field-level change in a proposal.
 */
export interface FieldChange {
  /** Node ID being changed */
  nodeId: NodeId;
  /** Field name */
  field: string;
  /** Old value */
  oldValue: unknown;
  /** New value */
  newValue: unknown;
}

/**
 * Result of field-level merge.
 */
export interface MergeResult {
  /** Successfully merged fields */
  merged: FieldChange[];
  /** Fields that conflict and need manual resolution */
  conflicts: {
    field: string;
    nodeId: NodeId;
    proposal1Value: unknown;
    proposal2Value: unknown;
  }[];
  /** Fields that were auto-merged */
  autoMerged: FieldChange[];
}

/**
 * Conflict detection result.
 */
export interface ConflictDetectionResult {
  /** Detected conflicts */
  conflicts: ProposalConflict[];
  /** Proposals that can be safely merged */
  mergeable: string[]; // Proposal IDs
  /** Proposals that need manual resolution */
  needsResolution: string[]; // Proposal IDs
}

/**
 * Conflict resolution strategy.
 */
export type ConflictStrategy =
  | "detect-only" // Just detect, don't auto-resolve
  | "field-level-merge" // Auto-merge non-conflicting fields
  | "manual-resolution" // Always require manual resolution
  | "last-write-wins" // Most recent approval wins
  | "first-write-wins" // First approval wins
  | "priority-based" // Use proposal priority
  | "supersede" // Use explicit superseding
  | "optimistic-lock"; // Reject stale proposals

/**
 * Configuration for conflict resolution.
 */
export interface ConflictConfiguration {
  /** Default strategy */
  defaultStrategy: ConflictStrategy;
  /** Per-node-type strategies */
  nodeTypeStrategies?: {
    [nodeType: string]: ConflictStrategy;
  };
  /** Whether to auto-merge field-level conflicts */
  autoMergeFields: boolean;
  /** Whether to require manual resolution for critical conflicts */
  requireManualResolution: boolean;
  /** Whether to use optimistic locking */
  useOptimisticLocking: boolean;
}

File: server/src/types/conflicts.rs
//! Conflict detection and merge types.
//! Per docs/core/AGENT_API.md § Conflict detection and merge and
//! docs/appendix/RECONCILIATION_STRATEGIES.md.

use serde::{Deserialize, Serialize};

use crate::types::NodeId;

/// One conflict between two proposals. Per AGENT_API: proposals, conflictingNodes,
/// conflictingFields?, severity, autoResolvable.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProposalConflict {
    /// Proposal IDs [id1, id2]
    pub proposals: Vec<String>,
    pub conflicting_nodes: Vec<NodeId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conflicting_fields: Option<std::collections::HashMap<String, Vec<String>>>,
    pub severity: ConflictSeverity,
    pub auto_resolvable: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConflictSeverity {
    Field,
    Node,
    Critical,
}

/// Result of detectConflicts(proposalId). Per AGENT_API: conflicts, mergeable, needsResolution.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConflictDetectionResult {
    pub conflicts: Vec<ProposalConflict>,
    pub mergeable: Vec<String>,
    pub needs_resolution: Vec<String>,
}

/// Field-level change. Used in MergeResult (merged, auto_merged).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldChange {
    pub node_id: NodeId,
    pub field: String,
    pub old_value: serde_json::Value,
    pub new_value: serde_json::Value,
}

/// One conflicting field in merge result. Per AGENT_API: field, nodeId, proposal1Value, proposal2Value.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MergeConflictField {
    pub field: String,
    pub node_id: NodeId,
    pub proposal1_value: serde_json::Value,
    pub proposal2_value: serde_json::Value,
}

/// Result of mergeProposals(proposalIds). Per AGENT_API: merged, conflicts, autoMerged.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MergeResult {
    pub merged: Vec<FieldChange>,
    pub conflicts: Vec<MergeConflictField>,
    pub auto_merged: Vec<FieldChange>,
}
